Node/src/models/o365ConnectorCard.ts

```diff
@@ -0,0 +1,577 @@
+        return this.data;
+    }
+}
+
+export class O365ConnectorCardFact implements teams.IIsO365ConnectorCardFact {
+    private data = <teams.IO365ConnectorCardFact>{ name: '' };
+    
+    constructor(private session?: builder.Session) {
+        
+    }
+    
+    public name(v: string): this {
+        this.data.name = v || '';
+        return this;
+    }
+
+    public value(text: string|string[], ...args: any[]): this {
+        if (text) {
+            this.data.value = fmtText(this.session, text, args);
+        }
+        return this;
+    }
+        
+    public toFact(): teams.IO365ConnectorCardFact {
+        return this.data;    
+    }
+}
+
+export class O365ConnectorCardImage implements teams.IIsO365ConnectorCardImage {
+    private data = <teams.IO365ConnectorCardImage>{};
+    
+    constructor(private session?: builder.Session) {
+        
+    }
+    
+    public image(url: string): this {
+        if (url) {
+            this.data.image = url;
+        }
+        return this;
+    }
+    
+    public title(text: string|string[], ...args: any[]): this {
+        if (text) {
+            this.data.title = fmtText(this.session, text, args);
+        }
+        return this;
+    }
+    
+    public toImage(): teams.IO365ConnectorCardImage {
+        return this.data;    
+    }
+}
+
+export abstract class O365ConnectorCardActionBase implements teams.IIsO365ConnectorCardActionBase {
+    protected data = <teams.IO365ConnectorCardActionBase>{}
+    
+    constructor(protected session?: builder.Session) {
+        this.data.type = this.type;
+    }
+
+    public name(text: string|string[], ...args: any[]): this {
+        if (text) {
+            this.data.name = fmtText(this.session, text, args);
+        }
+        return this;
+    }
+
+    public id(actionId: string): this {
+        if (actionId) {
+            this.data.id = actionId;
+        }
+        return this;
+    }
+
+    protected abstract get type(): string;
+
+    public toAction(): teams.IO365ConnectorCardActionBase {
+        return this.data;
+    }
+}
+
+export class O365ConnectorCardViewAction extends O365ConnectorCardActionBase {
+    constructor(protected session?: builder.Session) {
+        super(session);
+    }
+
+    public target(targetUrl: string): this {
+        if (targetUrl) {
+            (<teams.IO365ConnectorCardViewAction>this.data).target = [targetUrl];
+        }
+        return this;
+    }
+
+    protected get type(): string {
+        return 'ViewAction';
+    }
+}
+
+export class O365ConnectorCardOpenUri extends O365ConnectorCardActionBase {
+    private targets: teams.IO365ConnectorCardOpenUriTarget[] = [];
+
+    constructor(protected session?: builder.Session) {
+        super(session);
+        (<teams.IO365ConnectorCardOpenUri>this.data).targets = this.targets;
+    }
+
+    public default(targetUrl: string): this {
+        if (targetUrl) {
+            this.targets.push(<teams.IO365ConnectorCardOpenUriTarget> {
```

Comment:
I revised the implementation as update 2 where we guarantee data is always idempotent.